crates/subspace-transaction-pool/src/bundle_validator.rs

```diff
@@ -0,0 +1,256 @@
+}
+
+impl<Block: BlockT> BundleStoredInLastK<Block> {
+    fn new() -> Self {
+        BundleStoredInLastK {
+            bundle_stored_in_last_k: Arc::new(Mutex::new(VecDeque::new())),
+        }
+    }
+
+    fn update(&self, incoming: VecDeque<(Block::Hash, HashSet<Hash>)>) {
+        *self.bundle_stored_in_last_k.lock() = incoming;
+    }
+}
+
+pub struct BundleValidator<Block: BlockT, Client> {
+    client: Arc<Client>,
+    bundle_syncer: BundleStoredInLastK<Block>,
+}
+
+impl<Block: BlockT, Client> Clone for BundleValidator<Block, Client> {
+    fn clone(&self) -> Self {
+        BundleValidator {
+            client: self.client.clone(),
+            bundle_syncer: self.bundle_syncer.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct SkipBundleValidation;
+
+pub trait ValidateBundle<Block: BlockT, DomainHash: Encode> {
+    // for consensus chain, check the duplicated bundle and receipts.
+    // for system domain, checks nothing.
+    fn validate_bundle(
+        &self,
+        at: BlockId<Block>,
+        signed_opaque_bundle: &SignedOpaqueBundle<NumberFor<Block>, Block::Hash, DomainHash>,
+    ) -> sp_blockchain::Result<()>;
+}
+
+impl<Block: BlockT, DomainHash: Encode> ValidateBundle<Block, DomainHash> for SkipBundleValidation {
+    fn validate_bundle(
+        &self,
+        _at: BlockId<Block>,
+        _signed_opaque_bundle: &SignedOpaqueBundle<NumberFor<Block>, Block::Hash, DomainHash>,
+    ) -> sp_blockchain::Result<()> {
+        Ok(())
+    }
+}
+
+impl<Block, DomainHash, Client> ValidateBundle<Block, DomainHash> for BundleValidator<Block, Client>
+where
+    Block: BlockT,
+    DomainHash: Encode,
+    Client: HeaderBackend<Block>,
+{
+    fn validate_bundle(
+        &self,
+        at: BlockId<Block>,
+        signed_opaque_bundle: &SignedOpaqueBundle<NumberFor<Block>, Block::Hash, DomainHash>,
+    ) -> sp_blockchain::Result<()> {
+        // The hash used here must be the same as what is maintaining in `bundle_stored_in_last_k`,
+        // namely the hash of `SignedOpaqueBundle`
+        let incoming_bundle = signed_opaque_bundle.hash();
+        let block_bundles = self.bundle_syncer.bundle_stored_in_last_k.lock();
+        for (block_hash, bundle_hashes) in block_bundles.iter() {
+            // This implement will never return false negative result (i.e return `Err` for a new bundle)
+            // but it may return false positive result (i.e return `Ok` for a duplicated bundle) if
+            // `BundleHandler::on_block_import` return error and left some blocks unhandled, and it
+            // will be recovered after successfully handling the next best block.
+            if bundle_hashes.contains(&incoming_bundle) {
+                return Err(sp_blockchain::Error::Application(Box::from(format!(
+                    "Duplicated bundle {incoming_bundle}, bundle already included in block {block_hash}"
+                ))));
+            }
+        }
+
+        let block_number =
```

Comment:
better to call to it `best_primary_number` since this anyway assumes receipts are always constructed with number beloe best